src/scenic/core/specifiers.py

```diff
@@ -3,79 +3,119 @@
 import itertools
 
 from scenic.core.lazy_eval import (DelayedArgument, valueInContext, requiredProperties,
-                                   needsLazyEvaluation)
+                                   needsLazyEvaluation, toLazyValue)
 from scenic.core.distributions import toDistribution
-from scenic.core.errors import RuntimeParseError
+from scenic.core.errors import InvalidScenarioError, SpecifierError
 
 ## Specifiers themselves
 
 class Specifier:
-	"""Specifier providing a value for a property given dependencies.
-
-	Any optionally-specified properties are evaluated as attributes of the primary value.
-	"""
-	def __init__(self, prop, value, deps=None, optionals={}):
-		self.property = prop
-		self.value = value
-		if deps is None:
-			deps = set()
-		deps |= requiredProperties(value)
-		if prop in deps:
-			raise RuntimeParseError(f'specifier for property {prop} depends on itself')
-		self.requiredProperties = deps
-		self.optionals = optionals
-
-	def applyTo(self, obj, optionals):
-		"""Apply specifier to an object, including the specified optional properties."""
-		val = valueInContext(self.value, obj)
-		val = toDistribution(val)
-		assert not needsLazyEvaluation(val)
-		obj._specify(self.property, val)
-		for opt in optionals:
-			assert opt in self.optionals
-			obj._specify(opt, getattr(val, opt))
-
-	def __str__(self):
-		return f'<Specifier of {self.property}>'
+    """Specifier providing values for properties, at various priorities,
+    given dependencies.
+
+    Args:
+        name: The name of this specifier.
+        priorities: A dictionary mapping properties to the priority
+          they are being specified with.
+        value: A dictionary mapping properties to the values they are
+          being specified as.
+        deps: An iterable containing all properties that this
+          specifier relies on.
+    """
+    def __init__(self, name, priorities, value, deps=None):
+        assert isinstance(priorities, dict)
+        assert isinstance(value, (dict, DelayedArgument))
+
+        self.priorities = priorities
+        self.value = toLazyValue(value)
+
+        if deps is None:
+            deps = set()
+
+        deps |= requiredProperties(self.value)
+
+        for p in priorities:
+            if p in deps:
+                raise SpecifierError(f'specifier for property {p} depends on itself')
+
+        self.requiredProperties = tuple(sorted(deps))
+        self.name = name
+
+    def getValuesFor(self, obj):
+        """Get the values specified for a given object."""
+        val = valueInContext(self.value, obj)
+        assert isinstance(val, dict)
+        return val
+
+    def __str__(self):
+        return f'<{self.name} Specifier for {self.priorities}>'
+
+class ModifyingSpecifier(Specifier):
+    """Specifier providing values for properties, at various priorities,
+    given dependencies. Can also modify existing properties.
```

Comment:
Summary should be 1 line. Extra details should go in a new paragraph separated by a blank line.